### PR TITLE
Implement new PublishedEndpoints default in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR implements my PublishedEndpoints proposal (see #2806) in C#.

With this update, most application should:
 - configure their OA endpoints to be a single endpoint with a wildcard address
 - not set PublishedHost or PublishedEndpoints

For a follow-up PR:
 - simplify / cleanup the tests to listen on the wildcard address and take advantage of the new property (I wanted to demonstrate here that existing code - tests - continues to work unmodified)
 - add test for PublishedHost

### CODE REVIEW

File: csharp/src/Ice/ObjectAdapter.cs [hunk 1]
```diff
@@ -938,26 +939,12 @@ internal bool isLocal(Reference r)
...
+                foreach (EndpointI endpoint in endpoints)
                 {
-                    foreach (EndpointI endpoint in _publishedEndpoints)
-                    {
-                        if (endpoints[i].equivalent(endpoint))
-                        {
-                            return true;
-                        }
-                    }
-                    foreach (IncomingConnectionFactory factory in _incomingConnectionFactories)
```
Comment (bernardnormier) [inline csharp/src/Ice/ObjectAdapter.cs:955]: I simplified the "isLocal" check (used for collocation optimization) - with this update, we check only the published endpoints.

Previously, we checked the published endpoints and also the "published endpoint" of each incoming connection factory, which I find very much over the top.

File: csharp/src/Ice/Internal/IPEndpointI.cs [hunk 2]
```diff
@@ -150,21 +119,11 @@ public override List<EndpointI> expandHost(out EndpointI publishedEndpoint)
             instance_.preferIPv6(),
             true);
 
-        if (addresses.Count == 1)
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/IPEndpointI.cs:153]: This logic is not good, as we often end-up returning a DNS name (such as localhost) for the IncomingConnectionFactory endpoint. We do want actual IP addresses for these endpoints.

File: csharp/src/Ice/Internal/Network.cs [hunk 3]
```diff
@@ -803,7 +803,7 @@ public static List<EndPoint> getAddresses(
         return addresses;
     }
 
-    public static IPAddress[] getLocalAddresses(int protocol, bool includeLoopback, bool singleAddressPerInterface)
+    public static IPAddress[] getLocalAddresses(int protocol, bool singleAddressPerInterface)
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/Network.cs:806]: My other changes made the includeLoopback parameter always true, so I simplified this code.

File: csharp/src/Ice/Internal/Network.cs [hunk 4]
```diff
@@ -933,13 +933,12 @@ public static void
         }
     }
 
-    public static List<string> getHostsForEndpointExpand(string host, int protocol, bool includeLoopback)
+    public static List<string> getHostsForEndpointExpand(string host, int protocol)
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/Network.cs:935]: includeLoopback is (was) always true.

File: csharp/test/Ice/background/EndpointI.cs [hunk 5]
```diff
@@ -173,29 +173,8 @@ public EndpointI endpoint(Ice.Internal.EndpointI delEndp)
         }
     }
 
-    public override List<Ice.Internal.EndpointI> expandIfWildcard()
```
Comment (bernardnormier) [inline csharp/test/Ice/background/EndpointI.cs:176]: The update to this internal code is surprisingly the only update to the tests.

File: csharp/src/Ice/Internal/EndpointI.cs [hunk 6]
```diff
@@ -129,23 +129,15 @@ public virtual void streamWrite(Ice.OutputStream s)
     //
     public abstract Acceptor acceptor(string adapterName, SslServerAuthenticationOptions serverAuthenticationOptions);
 
-    //
-    // Expand endpoint out in to separate endpoints for each local
-    // host if listening on INADDR_ANY on server side or if no host
-    // was specified on client side.
-    //
-    public abstract List<EndpointI> expandIfWildcard();
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/EndpointI.cs:137]: We no longer need this expansion.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 7]
```diff
@@ -1406,57 +1388,53 @@ private List<EndpointI> parseEndpoints(string endpts, bool oaEndpoints)
...
         }
         else
         {
-            //
-            // Parse published endpoints. If set, these are used in proxies
-            // instead of the connection factory endpoints.
-            //
-            string endpts = _instance.initializationData().properties!.getProperty(_name + ".PublishedEndpoints");
+            // Parse published endpoints. If set, these are used instead of the connection factory endpoints.
+            string endpts = _instance.initializationData().properties!.getProperty($"{_name}.PublishedEndpoints");
```
Comment (bernardnormier) [inline csharp/src/Ice/ObjectAdapter.cs:1400]: I'll describe the algorithm on the issue.

File: csharp/test/IceSSL/configuration/AllTests.cs [hunk 8]
```diff
@@ -45,11 +45,13 @@ private static Dictionary<string, string>
     createServerProps(Ice.Properties defaultProperties)
     {
         Dictionary<string, string> result = new Dictionary<string, string>();
-        result["IceSSL.DefaultDir"] = defaultProperties.getProperty("IceSSL.DefaultDir");
-        result["Ice.Default.Host"] = defaultProperties.getProperty("Ice.Default.Host");
-        if (defaultProperties.getProperty("Ice.IPv6").Length > 0)
+        result["IceSSL.DefaultDir"] = defaultProperties.getIceProperty("IceSSL.DefaultDir");
```
Comment (bernardnormier) [inline csharp/test/IceSSL/configuration/AllTests.cs:48]: I find this conversion logic unexpected - and ideally we should fix it and just "convert" with:
https://github.com/zeroc-ice/ice/blob/2846e40a8ea07d870d762d89b2075071ab04aba2/csharp/src/Ice/Properties.cs#L285

This is however not the focus of this PR.

File: csharp/src/Ice/ObjectAdapter.cs (hunk 7 above)
Comment (pepone) [inline csharp/src/Ice/ObjectAdapter.cs:1400]: I would remove this variable and move this to the `parseEndpoints` call

```
endpoints = parseEndpoints(
    _instance.initializationData().properties!.getProperty($"{_name}.PublishedEndpoints"),
    oaEndpoints: false);
```

**Review by externl — APPROVED** [review]
Looks good to me!

File: csharp/test/IceSSL/configuration/PlatformTests.cs [hunk 9]
```diff
@@ -100,7 +100,9 @@ private static void clientRejectServerUsingValidationCallback(TestHelper helper,
         };
         using var clientCommunicator = createClient(clientOptions);
 
-        var obj = ServerPrxHelper.createProxy(clientCommunicator, "server:" + helper.getTestEndpoint(10, "ssl"));
+        var obj = ServerPrxHelper.createProxy(
+            clientCommunicator,
+            "server:" + helper.getTestEndpoint(10, "ssl") + " -h 127.0.0.1");
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/test/IceSSL/configuration/PlatformTests.cs:105]: That's from #2814.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 10]
```diff
@@ -938,26 +939,12 @@ internal bool isLocal(Reference r)
             {
                 checkForDeactivation();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs:940]: The general pattern is when we lock _mutex, we check for deactivation. There is no point is using a deactivated OA, and this check is "free" once we've lock _mutex.

The question here is "do we need to lock _mutex"? The answer is yes as long as we keep refreshPublishedEndpoints.

File: csharp/src/Ice/ObjectAdapter.cs (hunk 7 above)
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs:1400]: Fixed as suggested.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 11]
```diff
@@ -1406,57 +1388,53 @@ private List<EndpointI> parseEndpoints(string endpts, bool oaEndpoints)
...
+                if (publishedHost.Length > 0)
+                {
+                    // Replace the host in all endpoints by publishedHost (when applicable).
+                    endpoints = endpoints.Select(e => e.withPublishedHost(publishedHost)).Distinct();
+                }
+                // else keep the loopback-only endpoints as-is (with IP addresses)
             }
         }
 
-        if (_instance.traceLevels().network >= 1 && endpoints.Count > 0)
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs:1457]: Fixed as suggested.

File: csharp/test/IceSSL/configuration/PlatformTests.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/test/IceSSL/configuration/PlatformTests.cs]: Reverted update.

**Pull request merged**